local.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) The EfficientGo Authors.
+// Licensed under the Apache License 2.0.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Containerize inspects startFn and builds Go shim with local process endpoint that imports given `startFn` function.
+// Binary is then put in adhoc container and returned as runnable ready to be started.
+func Containerize(e Environment, name string, startFn func(context.Context) error) (Runnable, error) {
+	de, ok := e.(*DockerEnvironment)
+	if !ok {
+		return nil, errors.New("not implemented")
+	}
+
+	// Not portable, but good enough for local unit tests.
+	wd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+
+	strs := strings.Split(runtime.FuncForPC(reflect.ValueOf(startFn).Pointer()).Name(), ".")
+	funcName := strs[len(strs)-1]
+	pkg := strings.Join(strs[:len(strs)-1], ".")
+
+	modulePath := pkg
+	absModulePath := wd
+	for len(absModulePath) > 0 {
+		_, err := os.Stat(filepath.Join(absModulePath, "go.mod"))
+		if os.IsNotExist(err) {
+			absModulePath = filepath.Dir(absModulePath)
+			modulePath = filepath.Dir(modulePath)
+			continue
+		}
+		if err == nil {
+			break
+		}
+		return nil, err
+	}
+
+	if len(absModulePath) == 0 {
+		return nil, errors.Errorf("not a Go module %v", wd)
+	}
+
+	f := NewInstrumentedRunnable(e, name).WithPorts(map[string]int{"http": 80}, "http").Future()
+	dir := filepath.Join(f.Dir(), "shim")
+
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return nil, err
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, "go.mod"), []byte(fmt.Sprintf(goModTmpl, modulePath, modulePath, absModulePath)), os.ModePerm); err != nil {
+		return nil, err
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, "main.go"), []byte(fmt.Sprintf(mainFileTmpl, pkg, funcName)), os.ModePerm); err != nil {
+		return nil, err
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, "Dockerfile"), []byte(dockerFile), os.ModePerm); err != nil {
+		return nil, err
+	}
+
+	cmd := de.exec("go", "mod", "tidy")
+	cmd.Dir = dir
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return nil, errors.Wrap(err, string(out))
+	}
+
+	cmd = de.exec("go", "build", "-o", "exe", "main.go")
+	cmd.Dir = dir
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return nil, errors.Wrap(err, string(out))
+	}
+
+	imageTag := fmt.Sprintf("e2e-local-%v:dynamic", name)
+	cmd = de.exec("docker", "build", "-t", imageTag, ".")
+	cmd.Dir = dir
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return nil, errors.Wrap(err, string(out))
+	}
+	return f.Init(StartOptions{Image: imageTag}), nil
+}
+
+const (
+	dockerFile = `FROM ubuntu:21.10
+
+ADD ./exe /bin/exe
+
+ENTRYPOINT [ "/bin/exe" ]`
+
+	mainFileTmpl = `package main
```

Comment:
Maybe we can use https://github.com/dave/jennifer here later on! 🙂

Reply:
+1 on that :+1:; I'm also wondering if it wouldn't be better to add these constants into separate template files for better readability / more convenient access (if e.g. someone wants to edit the file and use different parent image for the container or make other changes to the 'base' `main.go` file).

Reply:
Hmm.. honestly I don't see a value in both Jennifer (what it improves?) and putting in the file (you don't have syntax highlight toolset anyway as we can give a `.go` suffix). Sounds like what we have is the simplest form? 🤔